cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1527,4 +1628,26 @@ private void emitTimeMetrics(String namespace, String appName, String metricName
     long timeTaken = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
     metricsCollectionService.getContext(tags).gauge(metricName, timeTaken);
   }
+
+  public void markAppsAsLatest(List<ApplicationId> appIds) throws Exception {
+    // Validate the appIds do not contain duplicates, i.e. we are not trying to mark
+    // multiple versions of the same application as latest
+    Set<String> validated = new HashSet<>();
+    for (ApplicationId appId: appIds) {
+      String appName = appId.getApplication();
+      if (validated.contains(appName)) {
+        throw new Exception(String.format(
+            "Marking multiple versions of an application (%s) as latest is not supported.",
+            appName
+        ));
+      }
+      validated.add(appName);
+    }
+
+    try {
+      store.markApplicationsLatest(appIds);
+    } catch (Exception e) {
+      throw new Exception("Failed to mark the applications as latest.");
```

Comment:
anytime an exception is wrapped, the cause *must* be specified, otherwise the root cause is lost. 

In this particular case, there is no reason to wrap the exception like this, the exception can be propagated up directly without catching.

Reply:
updated.